### PR TITLE
feat/SNAPWOO-6: Implement server-side Conversion Tracking 

### PR DESCRIPTION
**Pull request by Sidsector9**
Closes https://linear.app/a8c/issue/SNAPWOO-6/implement-server-side-tracking-using-snapchats-conversions-api-capi

## Description

- Implements server-side tracking via [CAPI](https://developers.snap.com/api/marketing-api/Conversions-API/GetStarted) for `ADD_CART` and `PURCHASE` events, and queues them using Action Scheduler jobs.
- Enables deduplication.
- Refactors Pixel tracking to be more modular.

## Testing

1. Visit the `/wp-admin/admin.php?page=wc-admin&path=%2Fsnapchat%2Fsettings` page.
2. Enable Conversions API tracking.
3. Add a product to the cart or make a purchase.
4. Visit the Action Scheduler page `/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending` and observe that the events get scheduled.
5. Verify if the jobs get executed successfully.

### CODE REVIEW

File: includes/Assets.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+/**
+ * Manages frontend asset loading for tracking features.
+ *
+ * Only enqueues assets if either pixel or conversion tracking is enabled.
+ * Uses {@see AssetLoader} for script registration and localization.
+ *
+ * @since 0.1.0
+ */
+class Assets {
```
Comment (joemcgill) [inline includes/Assets.php:34]: This should probably live in the Tracking namespace if it's specific to Tracking

File: includes/Assets.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			 *     Localized data for the frontend tracking script.
+			 *
+			 *     @type bool   $is_pixel_enabled      Whether pixel tracking is enabled.
+			 *     @type bool   $is_conversion_enabled Whether conversion tracking is enabled.
+			 *     @type string $capi_nonce            Nonce used for secure AJAX requests.
+			 * }
+			 *
+			 * @return array Modified tracking data.
+			 */
+			apply_filters(
```
Comment (joemcgill) [inline includes/Assets.php:99]: I'm curious why we're planning to use a filter to add additional tracking data, rather than including that data directly?

File: includes/Connection/WcsClient.php [hunk 3]
```diff
@@ -10,6 +10,7 @@
 
 namespace SnapchatForWooCommerce\Connection;
 
+use WpOrg\Requests\Requests;
```
Comment (joemcgill) [inline includes/Connection/WcsClient.php]: Are you planning to use the Requests library here for more than request type const values? Curious why this change was needed?

File: includes/Tracking/ConversionEvent/AddToCartEvent.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	/**
+	 * Builds the raw Conversion payload for the Ad Partner.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param array $args Overrideable payload args.
+	 *
+	 * @return array<string,mixed> Conversion event payload.
+	 */
+	public function build_payload( array $args = array() ): array {
```
Comment (joemcgill) [inline includes/Tracking/ConversionEvent/AddToCartEvent.php:60]: Since `event_id` is a requirement, we may want to add a check to make sure it's included and throw an error if not so the error will be more visible.

File: includes/Tracking/ConversionTrackingService.php [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param string $cart_item_key Unique key for the cart item.
+	 * @param int    $product_id    WooCommerce product ID added to the cart.
+	 * @param int    $quantity      Quantity of product added.
+	 * @return void
+	 */
+	public function handle_add_to_cart_old( string $cart_item_key, int $product_id, int $quantity ): void {
+		if ( Helper::is_add_to_cart_async() ) {
```
Comment (joemcgill) [inline includes/Tracking/ConversionTrackingService.php]: An inline comment to explain why this bails early here would be helpful.

File: includes/Tracking/ConversionTrackerInterface.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * These events are typically sent through an API endpoint provided by the Ad Partner,
+ * and may be transmitted immediately or via background job processors such as Action Scheduler.
+ *
+ * A matching client-side event (e.g. tracking pixel) is often sent in parallel to improve
+ * measurement accuracy. Deduplication identifiers (e.g., `event_id`) are used to avoid double attribution.
+ *
+ * @since 0.1.0
+ */
+interface ConversionTrackerInterface {
```
Comment (joemcgill) [inline includes/Tracking/ConversionTrackerInterface.php:31]: Should this specify a `send` method, since we're assuming one exists when registering the `send_conversion_event` callback?

File: includes/Tracking/ConversionTrackingService.php [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+	 * @since 0.1.0
+	 *
+	 * @param array $event_payload The payload that was sent to the Ad Partner.
+	 * @param array $args {
+	 *     Additional arguments passed to the tracking action.
+	 *
+	 *     @type int $order_id The ID of the WooCommerce order to mark as tracked.
+	 * }
+	 */
+	public function mark_as_tracked( $event_payload, $args ) {
```
Comment (joemcgill) [inline includes/Tracking/ConversionTrackingService.php:270]: Do we need to mark add_to_cart events similarly?

File: includes/Tracking/ConversionTrackingService.php [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+	 * @since 0.1.0
+	 *
+	 * @param string $cart_item_key Unique key for the cart item.
+	 * @param int    $product_id    WooCommerce product ID added to the cart.
+	 * @param int    $quantity      Quantity of product added.
+	 * @return void
+	 */
+	public function handle_single_product_add_to_cart( string $cart_item_key, int $product_id, int $quantity ): void {
+		/**
+		 * We only track synchronous Add to Cart events in this handler.
```
Comment (joemcgill) [inline includes/Tracking/ConversionTrackingService.php:197]: ✨ great use of inline docs here

File: includes/Tracking/RemotePixelTracker.php [hunk 9]
```diff
@@ -179,4 +185,104 @@ private function get_pixel_script() {
...
+					foreach ( $terms as $term ) {
+						$item_categories[] = $term->name;
+					}
+				}
+			}
+		}
+
+		$tracking_data = sprintf(
+			'snaptr("track", "PURCHASE", %s);',
+			wp_json_encode(
```
Comment (joemcgill) [inline includes/Tracking/RemotePixelTracker.php:287]: Doesn't this need to include the event_id for dedupe with CAPI?

File: includes/Tracking/RemotePixelTracker.php [hunk 10]
```diff
@@ -179,4 +185,104 @@ private function get_pixel_script() {
...
+
+
+	/**
+	 * Emits the Snapchat `VIEW_CONTENT` tracking event for single product views.
+	 *
+	 * Hooked into `woocommerce_after_single_product`.
+	 *
+	 * @since 0.1.0
+	 */
+	public function track_view_content_event(): void {
```
Comment (joemcgill) [inline includes/Tracking/RemotePixelTracker.php:203]: I see that VIEW_CONTENT, and PURCHASE events are moved here, but I'm not seeing an ADD_TO_CART event. Should that be handled here as well?

File: includes/Utils/AssetLoader.php
Comment (joemcgill) [inline includes/Utils/AssetLoader.php:1]: While not explicitly a part of this PR, we should consider calling `wp_enqueue_script` with a script loading strategy of `defer` rather than defaulting to printing it in the footer to speed up script loading slightly.

File: includes/Utils/Helper.php [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+	 * Check if the current request is an asynchronous "Add to Cart" action.
+	 *
+	 * This method determines whether the request is being made via AJAX
+	 * or through a REST API endpoint, both of which are considered asynchronous.
+	 *
+	 * @return bool True if the request is asynchronous (AJAX or REST), false otherwise.
+	 */
+	public static function is_add_to_cart_async() {
+		return ( wp_doing_ajax() || wp_is_serving_rest_request() );
+	}
```
Comment (joemcgill) [inline includes/Utils/Helper.php]: This utility might be better in the `ConversionTrackingService` unless you expect multiple services to need to make use of this utility. Alternately, the name should be more generic since this isn't specific to the add_to_cart action.

File: includes/Tracking/EventIdRegistry.php [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+	/**
+	 * Returns a unique event ID for the given product's add to cart event.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param int $product_id Product ID.
+	 * @return string Unique event ID.
+	 */
+	public static function get_add_to_cart_id( int $product_id ): string {
```
Comment (joemcgill) [inline includes/Tracking/EventIdRegistry.php]: I don't see this method in use anywhere. Is it planned for later?

File: src/tracking/conversions/utils.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+export const triggerCAPI = async ( eventId, productId, quantity ) => {
+	const formData = new FormData();
+	formData.append( 'action', TRACKING_DATA_VAR.capi_trigger_action );
+	formData.append( 'security', TRACKING_DATA_VAR.capi_nonce );
+	formData.append( 'product_id', productId );
+	formData.append( 'quantity', quantity );
+	formData.append( 'event_id', eventId );
+
+	await fetch( wc_add_to_cart_params.ajax_url, {
```
Comment (joemcgill) [inline src/tracking/conversions/utils.js]: Should probably add some validation handling in case `wc_add_to_cart_params` is undefined.

**Review by joemcgill — COMMENTED** [review]
Nice work on this @Sidsector9. Left a round of comments, questions, and suggestions.

File: includes/Assets.php (hunk 1 above)
Comment (Sidsector9) [reply includes/Assets.php:34]: It is specific to Tracking now, but as we integrate more features and services down the line, it would make sense to keep this outside as a common entry point to load assets required by multiple services.

File: includes/Connection/WcsClient.php (hunk 3 above)
Comment (Sidsector9) [reply includes/Connection/WcsClient.php]: You're right, it'll autoload the whole library. I've reverted.

File: includes/Tracking/ConversionTrackingService.php (hunk 7 above)
Comment (Sidsector9) [reply includes/Tracking/ConversionTrackingService.php:270]: No, a Purchase event happens once per order, but Add to Cart can happen multiple times for the same product before placing an order.
We mark orders because we don't want to send PURCHASE events when the Thank you page is refreshed/reloaded.

File: includes/Tracking/RemotePixelTracker.php (hunk 9 above)
Comment (Sidsector9) [reply includes/Tracking/RemotePixelTracker.php:287]: For `PURCHASE` events, Pixel tracking uses the `transaction_id` field, and CAPI uses `event_id`.
For all non-purchase events, Pixel tracking uses the `client_dedup_id` field and CAPI uses `event_id`.

I have validated the payload using [this artilcle](https://businesshelp.snapchat.com/s/article/capi-event-testing?language=en_US).

File: includes/Tracking/RemotePixelTracker.php (hunk 10 above)
Comment (Sidsector9) [reply includes/Tracking/RemotePixelTracker.php:203]: No, `VIEW_CONTENT` and `PURCHASE` pixel events are injected during page render time as we want to trigger them on page load. Whereas, `ADD_TO_CART` is implemented and triggered on the client-side when the button is clicked.

File: includes/Tracking/EventIdRegistry.php (hunk 12 above)
Comment (Sidsector9) [reply includes/Tracking/EventIdRegistry.php]: I missed deleting that out when I refactored the architecture.

File: includes/Tracking/ConversionEvent/AddToCartEvent.php (hunk 4 above)
Comment (Sidsector9) [reply includes/Tracking/ConversionEvent/AddToCartEvent.php:60]: It is _recommended_ at best if both Pixel and Conversion are enabled.

File: includes/Assets.php (hunk 2 above)
Comment (Sidsector9) [reply includes/Assets.php:99]: We are tracking CAPI `PURCHASE` event and marking only when the background job is completed. It's more convenient to use an action hook for this specific case.

File: includes/Assets.php (hunk 1 above)
Comment (joemcgill) [reply includes/Assets.php:34]: Got it. If the intent is for this to be the general class responsible for enqueueing assets, then it would be helpful to clarify the inline docs a bit.

File: includes/Tracking/RemotePixelTracker.php (hunk 10 above)
Comment (joemcgill) [reply includes/Tracking/RemotePixelTracker.php:203]: Thanks for following up. If ADD_TO_CART is only ever triggered client side then I think this is good.

File: includes/Tracking/RemotePixelTracker.php (hunk 9 above)
Comment (joemcgill) [reply includes/Tracking/RemotePixelTracker.php:287]: I think I have a different understanding, based on [these docs](https://developers.snap.com/api/marketing-api/Conversions-API/UsingTheAPI#deduplication), of how the `event_id` is meant to be used when deduping the same event that gets tracked by Pixel and by CAPI.

While the docs there suggest mapping the `event_id` to `client_dedup_id` (used for [client side deduplication](https://businesshelp.snapchat.com/s/article/deduplication-pixel?language=en_US)) and `transaction_id` values, I believe the intent is that events sent to all integrations (e.g. CAPI and Pixel) would all include an `event_id` value, which is used to match and dedupe cases where the same event (e.g., a purchase) is sent via a Pixel AND a CAPI request event.

File: includes/Utils/AssetLoader.php
Comment (joemcgill) [reply includes/Utils/AssetLoader.php:1]: Created a new issue in the backlog for this (SNAPWOO-16) so we can revisit in the future.

File: includes/Assets.php [hunk 14]
```diff
@@ -90,6 +90,7 @@ public function enqueue_scripts() {
 			apply_filters(
 				Helper::with_prefix( 'filter_tracking_data' ),
 				array(
+					'ajax_url'              => admin_url( 'admin-ajax.php' ),
```
Comment (joemcgill) [inline includes/Assets.php:102]: 👍🏻

**Review by joemcgill — COMMENTED** [review]
@Sidsector9 I still have some concerns about the need to send an `event_id` property to pixel events, which I've detailed in [this thread](https://github.com/woocommerce/snapchat-for-woocommerce/pull/3#discussion_r2167575254).

Once we're aligned there, I think this is ready to merge.

File: includes/Tracking/RemotePixelTracker.php (hunk 9 above)
Comment (Sidsector9) [reply includes/Tracking/RemotePixelTracker.php:287]: okay I understood what you mean. I fixed this in [8eb0f32](https://github.com/woocommerce/snapchat-for-woocommerce/pull/3/commits/8eb0f32d2c0fd9d3fde0df601ad3072321414ba2).

**Review by joemcgill — APPROVED** [review]
Thanks, @Sidsector9!
